File: frontend/src/features/workspace/lib/model-roles.ts
import type { ModelOption, ModelRole, ModelSelections, RoleModels } from "@/lib/api";

/**
 * 모델 용도의 표시 문구. 서버는 role 키만 내려주므로 문구는 프론트가 가집니다.
 * `unavailable`은 이 키로 쓸 수 있는 모델이 하나도 없을 때 무엇이 안 되는지 알려줍니다.
 */
export const modelRoleInfo: Record<
  ModelRole,
  { label: string; description: string; unavailable: string }
> = {
  answer: {
    label: "답변 생성",
    description: "Ask 화면에서 질문에 답하는 모델입니다.",
    unavailable: "질문에 답할 수 없습니다.",
  },
  extraction: {
    label: "분석·그래프 추출",
    description: "회의와 문서에서 결정·이슈·담당자를 뽑아 그래프와 현재 상황을 만드는 모델입니다.",
    unavailable: "결정·담당자·관계를 뽑아 그래프와 현재 상황을 만들 수 없습니다.",
  },
  embedding: {
    label: "검색 임베딩",
    description: "문서와 회의 내용을 검색할 수 있게 벡터로 바꾸는 모델입니다.",
    unavailable: "문서와 회의 내용을 검색할 수 없습니다.",
  },
  transcription: {
    label: "녹음 받아쓰기",
    description:
      "회의 녹음 파일을 글로 옮기는 모델입니다. 브라우저 받아쓰기에는 필요하지 않습니다.",
    unavailable: "녹음 파일은 받아쓸 수 없습니다. 브라우저 받아쓰기 대본은 그대로 쓸 수 있습니다.",
  },
};

export function optionKey(option: ModelOption): string {
  return `${option.provider}/${option.model}`;
}

/** 저장된 선택이 있으면 그것을, 없으면 서버가 추천한 첫 옵션을 미리 고릅니다. */
export function initialSelections(roles: RoleModels[]): ModelSelections {
  const selections: ModelSelections = {};
  for (const entry of roles) {
    const chosen = entry.selected ?? entry.options[0];
    if (chosen) selections[entry.role] = chosen;
  }
  return selections;
}

/** 두 선택이 같은지. 설정 화면에서 바뀐 용도만 저장하려고 비교합니다. */
export function sameSelection(a: ModelOption | undefined, b: ModelOption | undefined): boolean {
  return a?.provider === b?.provider && a?.model === b?.model;
}

File: backend/app/modules/context_engine/application/model_roles.py
"""Which models a key can use for which job.

The server never picks a model by itself: it asks the provider what the key can use, sorts the
answer into jobs (roles), and the user chooses. Whether a workspace can, say, embed is therefore a
fact about the models its key offers, not about the provider's name.
"""

import re
from datetime import date
from enum import StrEnum

from pydantic import BaseModel

from app.modules.context_engine.application.provider import ModelInfo, ProviderAdapter


class ModelRole(StrEnum):
    ANSWER = "answer"
    EXTRACTION = "extraction"
    EMBEDDING = "embedding"
    TRANSCRIPTION = "transcription"


ROLE_ORDER = (
    ModelRole.ANSWER,
    ModelRole.EXTRACTION,
    ModelRole.EMBEDDING,
    ModelRole.TRANSCRIPTION,
)

# What the provider's API must be able to do for a model to serve the role at all.
ROLE_CAPABILITY = {
    ModelRole.ANSWER: "chat",
    ModelRole.EXTRACTION: "structuredOutput",
    ModelRole.EMBEDDING: "embedding",
    ModelRole.TRANSCRIPTION: "transcription",
}

# Model ids that are not chat models even though a provider lists them.
_NOT_CHAT = re.compile(
    r"embed|whisper|transcribe|tts|dall-e|image|moderation|realtime|audio|babbage|davinci|search"
)
_DATED = re.compile(r"-(\d{4}-\d{2}-\d{2}|\d{8}|\d{4})$")
# Lighter (cheaper, faster) models come first: the choice is the user's, but what is preselected
# should not be the most expensive model of a family.
_LIGHT = re.compile(r"mini|nano|small|haiku|flash|lite")


class ModelOption(BaseModel):
    provider: str
    model: str


def roles_for_model(model_id: str, capabilities: tuple[str, ...]) -> set[ModelRole]:
    """The jobs `model_id` can do, judged by its id and by what the provider's API supports."""
    name = model_id.lower()
    roles: set[ModelRole] = set()
    if "embed" in name:
        roles.add(ModelRole.EMBEDDING)
    elif "whisper" in name or "transcribe" in name:
        roles.add(ModelRole.TRANSCRIPTION)
    elif not _NOT_CHAT.search(name):
        roles |= {ModelRole.ANSWER, ModelRole.EXTRACTION}
    return {role for role in roles if ROLE_CAPABILITY[role] in capabilities}


def recommendation_rank(model_id: str) -> tuple[bool, bool, str]:
    """Order of preselection: lighter models first, a stable alias before a dated snapshot.

    This is only a stable, explainable ordering. It says nothing about quality; the user picks.
    """
    return (not _LIGHT.search(model_id.lower()), bool(_DATED.search(model_id)), model_id)


def options_by_role(
    listings: list[tuple[ProviderAdapter, list[ModelInfo]]],
    prices: dict[tuple[str, str], float] | None = None,
    *,
    defaults: dict[str, dict[str, str]] | None = None,
    today: date | None = None,
) -> dict[ModelRole, list[ModelOption]]:
    """Exclude retired models, put offered defaults first, then sort other models by price."""
    today = today or date.today()
    options: dict[ModelRole, list[ModelOption]] = {role: [] for role in ROLE_ORDER}
    for adapter, infos in listings:
        live = {i.id for i in infos if i.shutdown_date is None or i.shutdown_date > today}
        preferred = (defaults or {}).get(adapter.id, {})
        for model in sorted(live, key=recommendation_rank):
            for role in ROLE_ORDER:
                if role in roles_for_model(model, adapter.capabilities):
                    option = ModelOption(provider=adapter.id, model=model)
                    if option in options[role]:
                        continue
                    if preferred.get(role.value) == model:
                        # The default leads its provider's models, after earlier keys' models.
                        first_of_provider = next(
                            (i for i, o in enumerate(options[role]) if o.provider == adapter.id),
                            len(options[role]),
                        )
                        options[role].insert(first_of_provider, option)
                    else:
                        options[role].append(option)
    if prices:
        for role in ROLE_ORDER:
            options[role].sort(
                key=lambda option: (
                    (defaults or {}).get(option.provider, {}).get(role.value) != option.model,
                    (option.provider, option.model) not in prices,
                    prices.get((option.provider, option.model), float("inf")),
                    recommendation_rank(option.model),
                    option.provider,
                )
            )
    return options


def selection_of(model_settings: dict[str, object] | None, role: ModelRole) -> ModelOption | None:
    """The workspace's stored choice for `role`, or None if it never chose."""
    raw = (model_settings or {}).get(role.value)
    if not isinstance(raw, dict):
        return None
    try:
        return ModelOption.model_validate(raw)
    except ValueError:
        return None


def invalid_selections(
    selections: dict[str, ModelOption], options: dict[ModelRole, list[ModelOption]]
) -> list[str]:
    """Why a set of choices is not acceptable: unknown role, or a model the key does not offer."""
    problems: list[str] = []
    for name, choice in selections.items():
        try:
            role = ModelRole(name)
        except ValueError:
            problems.append(f"알 수 없는 용도입니다: {name}")
            continue
        if choice not in options[role]:
            problems.append(
                f"{choice.provider}/{choice.model}은(는) {name} 용도로 쓸 수 없는 모델입니다."
            )
    return problems
